mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py

```diff
@@ -0,0 +1,203 @@
+                max_scores, _ = nms_pre_score[..., :-1].max(-1)
+            _, topk_inds = max_scores.topk(pre_topk)
+            batch_inds = torch.arange(
+                batch_size,
+                device=bbox_pred.device).view(-1, 1).expand_as(topk_inds)
+            priors = priors[batch_inds, topk_inds, :]
+            bbox_pred = bbox_pred[batch_inds, topk_inds, :]
+            scores = scores[batch_inds, topk_inds, :]
+            if with_score_factors:
+                score_factors = score_factors[batch_inds, topk_inds, :]
+
+        mlvl_valid_bboxes.append(bbox_pred)
+        mlvl_valid_scores.append(scores)
+        priors = self.anchor_center(priors)
+        mlvl_valid_priors.append(priors)
+        if with_score_factors:
+            mlvl_score_factors.append(score_factors)
+
+    batch_mlvl_bboxes_pred = torch.cat(mlvl_valid_bboxes, dim=1)
+    batch_scores = torch.cat(mlvl_valid_scores, dim=1)
+    batch_priors = torch.cat(mlvl_valid_priors, dim=1)
+    batch_bboxes = self.bbox_coder.decode(
+        batch_priors, batch_mlvl_bboxes_pred, max_shape=img_shape)
+    if with_score_factors:
+        batch_score_factors = torch.cat(mlvl_score_factors, dim=1)
+
+    if not self.use_sigmoid_cls:
+        batch_scores = batch_scores[..., :self.num_classes]
+
+    if with_score_factors:
+        batch_scores = batch_scores * batch_score_factors
+    if not with_nms:
+        return batch_bboxes, batch_scores
+    post_params = get_post_processing_params(deploy_cfg)
+    max_output_boxes_per_class = post_params.max_output_boxes_per_class
+    iou_threshold = cfg.nms.get('iou_threshold', post_params.iou_threshold)
+    score_threshold = cfg.get('score_thr', post_params.score_threshold)
+    pre_top_k = post_params.pre_top_k
+    keep_top_k = cfg.get('max_per_img', post_params.keep_top_k)
+    return multiclass_nms(
+        batch_bboxes,
+        batch_scores,
+        max_output_boxes_per_class,
+        iou_threshold=iou_threshold,
+        score_threshold=score_threshold,
+        pre_top_k=pre_top_k,
+        keep_top_k=keep_top_k)
+
+
+def batched_integral(intergral, x):
+    batch_size = x.size(0)
+    x = F.softmax(x.reshape(batch_size, -1, intergral.reg_max + 1), dim=2)
+    x = F.linear(x,
+                 intergral.project.type_as(x).unsqueeze(0)).reshape(
+                     batch_size, -1, 4)
+    return x
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    func_name='mmdet.models.dense_heads.gfl_head.'
+    'GFLHead.forward_single')
+def gfl_head__forward_single(ctx, self, x, scale, **kwargs):
+    """Rewrite `forward_single` of `GFLHead` for default backend.
+
+    Rewrite this function to deploy model, (by richard.)
+    """
+    cls_feat = x
+    reg_feat = x
+    for cls_conv in self.cls_convs:
+        cls_feat = cls_conv(cls_feat)
+    for reg_conv in self.reg_convs:
+        reg_feat = reg_conv(reg_feat)
+    cls_score = self.gfl_cls(cls_feat)
+    bbox_pred = scale(self.gfl_reg(reg_feat)).float().permute(
```

Comment:
Could permute in `gfl_head__get_bbox` before `batched_integral`. So we can remove `gfl_head__forward_single`.

Reply:
I hope rewrite does not change the behavior of forward. We might reuse any module in mmdetection(or other repo), if we change the behavior of forward, other head which share the same forward(if it did exist) might get unexpect result.

Reply:
OK, I see what you mean, thank you.